tensorflow/python/keras/mixed_precision/experimental/autocast_variable.py

```diff
@@ -188,61 +188,88 @@ def initial_value(self):
   def constraint(self):
     return self._variable.constraint
 
+  def _apply_assign_update(
+      self, update_fn, value, use_locking=None, name=None, read_value=True):
+    if not read_value:
+      return update_fn(value, use_locking, name, read_value)
+
+    if context.executing_eagerly() or ops.inside_function():
+      assign_op = update_fn(value, use_locking, name, False)
+      with ops.control_dependencies([assign_op]):
```

Comment:
This control dependency has no effect. It only affects ops created under it, but there are no ops created. Luckily, I don't think it matters since tf.functions automatically create control dependencies to ensure stateful ops are executed.

And same for `_apply_update`

Reply:
Also I think we want to return `None` here if `read_value` is False, as is done by `tf.Variable`.

Reply:
> Luckily, I don't think it matters since tf.functions automatically create control dependencies to ensure stateful ops are executed.

:+1: Good point. I wasn't sure about that since `ops.control_dependencies` is used quite a lot inside TensorFlow but I guess this is only for compatibility reasons with legacy v1 graph mode then.

> Also I think we want to return `None` here if `read_value` is `False`, as is done by `tf.Variable`.

Looks like `tf.Variable.assign` might return an operations inside tf functions:
```python
x = tf.Variable(1.0)

@tf.function
def foo():
    print(type(x.assign(2.0, read_value=False)))

foo()  # <class 'tensorflow.python.framework.ops.Operation'>
```
But without the need for control dependencies it still can be simplified to `return self if read_value else assign_op`.